src/internet_identity/src/archive.rs

```diff
@@ -0,0 +1,179 @@
+use crate::state;
+use crate::state::{ArchiveData, ArchiveState, DeviceDataInternal};
+use candid::Principal;
+use ic_cdk::api::call::CallResult;
+use ic_cdk::api::management_canister::main::{
+    canister_status, create_canister, install_code, CanisterIdRecord, CanisterInstallMode,
+    CanisterInstallMode::Install, CreateCanisterArgument, InstallCodeArgument,
+};
+use ic_cdk::api::time;
+use ic_cdk::{id, notify};
+use internet_identity_interface::{
+    ArchiveInit, DeployArchiveResult,
+    DeployArchiveResult::{CreationFailed, UpgradeFailed},
+    DeviceDataUpdate, Entry, Operation, Private, UserNumber,
+};
+use serde_bytes::ByteBuf;
+use sha2::Digest;
+use sha2::Sha256;
+use CanisterInstallMode::Upgrade;
+
+pub async fn deploy_archive(wasm: ByteBuf) -> DeployArchiveResult {
+    let archive_state = state::persistent_state_mut(|persistent_state| {
+        let archive_state = persistent_state.archive_info.state.clone();
+        if archive_state == ArchiveState::NotCreated {
+            // lock archive creation because of async operation
+            persistent_state.archive_info.state = ArchiveState::CreationInProgress;
+        }
+        archive_state
+    });
+
+    match archive_state {
+        ArchiveState::NotCreated => create_and_install_archive(wasm.into_vec()).await,
+        ArchiveState::CreationInProgress => DeployArchiveResult::CreationInProgress,
```

Comment:
just confirming my assumption about semantics here: if `CreationInProgress` is returned, it means that _some other call_ triggered a deploy, and there is no particular reason why the wasm module being deployed is the same as `wasm`, correct?

Reply:
> just confirming my assumption about semantics here: if `CreationInProgress` is returned, it means that _some other call_ triggered a deploy, and there is no particular reason why the wasm module being deployed is the same as `wasm`, correct?

Yes, exactly. Also note, that the lock on archive operation is during archive creation _only_, to avoid creating multiple archive canisters due to race conditions.

But if you were to upgrade II in rapid succession to change the `wasm` hash for the archive and deploy then deploy with different hashes, the order in which the deploy messages get executed is up to the platform. Although practically, I don't think you can upgrade II fast enough, given the archive and II are on the same subnet, so that the deploy messages could cross and get reordered.